airflow-core/src/airflow/dag_processing/processor.py

```diff
@@ -234,6 +238,39 @@ def _serialize_dags(
     return serialized_dags, serialization_import_errors
 
 
+def _get_dag_with_task(
+    dagbag: DagBag, dag_id: str, task_id: str | None = None
+) -> tuple[DAG, BaseOperator | MappedOperator | None]:
+    """
+    Retrieve a DAG and optionally a task from the DagBag.
+
+    :param dagbag: DagBag to retrieve from
+    :param dag_id: DAG ID to retrieve
+    :param task_id: Optional task ID to retrieve from the DAG
+    :return: tuple of (dag, task) where task is None if not requested
+    :raises ValueError: If DAG or task is not found
+    """
+    if dag_id not in dagbag.dags:
+        raise ValueError(
+            f"DAG '{dag_id}' not found in DagBag. "
+            f"This typically indicates a race condition where the DAG was removed or failed to parse."
+        )
+
+    dag = dagbag.dags[dag_id]
+
+    if task_id is not None:
+        try:
+            task = dag.get_task(task_id)
+            return dag, task
+        except TaskNotFound:
+            raise ValueError(
+                f"Task '{task_id}' not found in DAG '{dag_id}'. "
+                f"This typically indicates a race condition where the task was removed or the DAG structure changed."
```

Comment:
@kaxil , just trying to understand why we are raising `ValueError` instead of `DagNotFound` or `TaskNotFound` here?

Reply:
Ideally we shouldn't need those special exceptions for things like this:

https://lists.apache.org/thread/t8bnhyqy77kq4fk7fj3fmjd5wo9kv6w0

Reply:
I agree about not using `AirflowException`. But, user defined exceptions can be helpful to maintain consistent error messages. And, I think, we should still have `DagNotFoundException`. But, this should just extend `Exception` instead of `AirflowException`. Wdyt?

Reply:
at least the error messages that User see's, they should be consistent and we can maintain such consistency by using such user defined exceptions.